xla/pjrt/pjrt_stream_executor_client.cc

```diff
@@ -3739,6 +3739,35 @@ PjRtStreamExecutorClient::LoadInternal(
   const bool xla_gpu_dump_hlo_unoptimized_snapshots =
       ex_options.has_debug_options() &&
       ex_options.debug_options().xla_gpu_dump_hlo_unoptimized_snapshots();
+  // Override debug_options() with those explicitly passed in when
+  // deserializing. This allows options such as --xla_dump_to to be changed.
+  if (ex_options.has_debug_options()) {
+    for (std::unique_ptr<LocalExecutable>& local_executable :
+         local_executables) {
+      HloModule& hlo_module = local_executable->executable()->module();
+      if (std::string serialized_only_values = GetNonDefaultDebugOptions(
+              hlo_module.config().debug_options(), ex_options.debug_options());
+          !serialized_only_values.empty()) {
+        VLOG(1) << "Replacing serialized debug options:\n"
+                << serialized_only_values << "with passed debug options:\n"
+                << GetNonDefaultDebugOptions(
+                       ex_options.debug_options(),
+                       hlo_module.config().debug_options());
+      }
+      hlo_module.mutable_config().mutable_debug_options() =
+          ex_options.debug_options();
+    }
+  }
```

Comment:
I think this is too dangerous. There are probably various implicit dependencies between the runtime, the debug options, and the HLO module. Changing arbitrary DebugOptions might break things in unexpected ways.

I think a better way to do this is to change `DumpHloModuleIfEnabled` in a way that the DebugOptions can be supplied as a separate argument.

Reply:
OK, I tried to plumb through the relevant options more explicitly.